File: No_1019/Next Greater Node In Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<int> nextLargerNodes(ListNode* head) {
        stack<pair<int, int>> s;
        vector<int> ret;
        int index = 0;
        
        while(head != nullptr) {
            ret.push_back(0);
            while(!s.empty() && s.top().first < head->val) {
                ret[s.top().second] = head->val;
                s.pop();
            }
        
            s.push(make_pair(head->val, index));
            head = head->next;
            index++;
        }
        
        return ret;
    }
};

File: No_1019/Next Greater Node In Linked List.py
class Solution:
    def nextLargerNodes(self, head: Optional[ListNode]) -> List[int]:
        ans = []
        stack = []  # (val, index)
        index = 0

        while head:
            ans.append(0)
            while stack and head.val > stack[-1][0]:
                _, idx = stack.pop()
                ans[idx] = head.val
            stack.append((head.val, index))
            index += 1
            head = head.next

        return ans